File: 0155_Min_Stack/Solution.java
import java.util.ArrayList;


class Item {
    int val;
    int min;

    public Item(int val, int min) {
        this.val = val;
        this.min = min;
    }
}


class MinStack {
    private ArrayList<Item> content;

    public MinStack() {
        content = new ArrayList<>();
    }

    public void push(int val) {
        if (content.isEmpty()) {
            content.add(new Item(val, val));
        } else {
            int min = Math.min(val, content.getLast().min);
            content.add(new Item(val, min));
        }
    }

    public void pop() {
        if (!content.isEmpty()) {
            content.removeLast();
        }
    }

    public int top() {
        if (!content.isEmpty()) {
            return content.getLast().val;
        }
        throw new IllegalStateException("Stack is empty");
    }

    public int getMin() {
        if (!content.isEmpty()) {
            return content.getLast().min;
        }
        throw new IllegalStateException("Stack is empty");
    }
}

File: 0155_Min_Stack/solution.py
type Item = tuple[int, int]

class MinStack:
    
    def __init__(self):
        self.content: list[Item] = []

    def push(self, val: int):
        if not self.content:
            self.content.append((val, val))
        else:
            _, prev_min = self.content[-1]
            self.content.append((val, min(val, prev_min)))

    def pop(self):
        self.content.pop()

    def top(self) -> int:
        return self.content[-1][0]

    def get_min(self) -> int:
        return self.content[-1][1]